go/cli/mcap/utils/mcap_amendment.go

```diff
@@ -0,0 +1,454 @@
+package utils
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"hash/crc32"
+	"io"
+
+	"github.com/foxglove/mcap/go/mcap"
+)
+
+const (
+	SIZE_RECORD_LENGTH = 8
+	SIZE_OPCODE        = 1
+	SIZE_DATA_END      = 4
+	SIZE_FOOTER        = 8 + 8 + 4
+	SIZE_MAGIC         = 8
+)
+
+// AmendMCAP adds attachment and metadata records to the end of the
+// data section of an existing MCAP, in place. It then writes a new summary
+// section consisting of the existing summary section, plus new attachment and
+// metadata index records as applicable, with all offsets and CRCs updated to
+// account for the new data.
+func AmendMCAP(
+	wsc io.ReadWriteSeeker,
+	attachments []*mcap.Attachment,
+	metadata []*mcap.Metadata,
+) error {
+	lexer, err := mcap.NewLexer(wsc, &mcap.LexerOptions{
+		SkipMagic: true,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to build lexer: %w", err)
+	}
+	defer lexer.Close()
+
+	oldFooter, err := readFooter(wsc)
+	if err != nil {
+		return fmt.Errorf("failed to read footer: %w", err)
+	}
+	oldSummaryStart := oldFooter.SummaryStart
+
+	// seek to the opcode of the data end record
+	oldDataEndOffset := int64(oldSummaryStart - SIZE_DATA_END - SIZE_RECORD_LENGTH - SIZE_OPCODE)
+	oldDataEnd, err := readDataEnd(wsc, oldDataEndOffset)
+	if err != nil {
+		return fmt.Errorf("failed to read data end: %w", err)
+	}
+
+	// Now we are at the old summary start. Parse from here to EOF into a
+	// summarySection structure.
+	summary, err := readSummarySection(wsc)
+	if err != nil {
+		return fmt.Errorf("failed to read summary section: %w", err)
+	}
+
+	// Seek to the location of the old data end record. We will write the new data section records starting here.
+	_, err = wsc.Seek(oldDataEndOffset, io.SeekStart)
+	if err != nil {
+		return fmt.Errorf("failed to seek to data end: %w", err)
+	}
+
+	extraDataLength, attachmentIndexes, metadataIndexes, err := writeExtraData(
```

Comment:
I might have named this `extendDataSection`, which feels more like it would add a DataEnd record than `writeExtraData`. I defer to your taste though.